// svelte/src/internal/client/errors.js
function effect_in_teardown(rune) {
  {
    throw new Error("effect_in_teardown");
  }
}
function effect_in_unowned_derived() {
  {
    throw new Error("effect_in_unowned_derived");
  }
}
function effect_orphan(rune) {
  {
    throw new Error("effect_orphan");
  }
}
function effect_update_depth_exceeded() {
  {
    throw new Error("effect_update_depth_exceeded");
  }
}
function props_invalid_value(key) {
  {
    throw new Error("props_invalid_value");
  }
}
function state_descriptors_fixed() {
  {
    throw new Error("state_descriptors_fixed");
  }
}
function state_prototype_fixed() {
  {
    throw new Error("state_prototype_fixed");
  }
}
function state_unsafe_local_read() {
  {
    throw new Error("state_unsafe_local_read");
  }
}
function state_unsafe_mutation() {
  {
    throw new Error("state_unsafe_mutation");
  }
}
export {
  effect_in_teardown,
  effect_in_unowned_derived,
  effect_orphan,
  effect_update_depth_exceeded,
  props_invalid_value,
  state_descriptors_fixed,
  state_prototype_fixed,
  state_unsafe_local_read,
  state_unsafe_mutation
};
